compiler/noirc_frontend/src/hir_def/types.rs

```diff
@@ -1485,16 +1485,28 @@ impl Type {
     /// is used and generic substitutions are provided manually by users.
     ///
     /// Expects the given type vector to be the same length as the Forall type variables.
-    pub fn instantiate_with(&self, types: Vec<Type>) -> (Type, TypeBindings) {
+    pub fn instantiate_with(
+        &self,
+        types: Vec<Type>,
+        interner: &NodeInterner,
+        implicit_generic_count: usize,
+    ) -> (Type, TypeBindings) {
         match self {
             Type::Forall(typevars, typ) => {
-                assert_eq!(types.len(), typevars.len(), "Turbofish operator used with incorrect generic count which was not caught by name resolution");
+                assert_eq!(types.len() + implicit_generic_count, typevars.len(), "Turbofish operator used with incorrect generic count which was not caught by name resolution");
 
-                let replacements = typevars
+                let mut replacements = typevars[..implicit_generic_count]
                     .iter()
-                    .zip(types)
-                    .map(|(var, binding)| (var.id(), (var.clone(), binding)))
-                    .collect();
+                    .map(|var| {
+                        let new = interner.next_type_variable();
+                        (var.id(), (var.clone(), new))
+                    })
+                    .collect::<HashMap<_, _>>();
+
+                // `zip` ends at the shorter iterator so we skip any implicit generics included in the `Forall` type
+                for (var, binding) in typevars.iter().skip(implicit_generic_count).zip(types) {
+                    replacements.insert(var.id(), (var.clone(), binding));
+                }
```

Comment:
You should only need one of these loops. The first inserts fresh type variables while the second inserts concrete types. The way you have it also skips over instantiating implicit generics at all, which we don't want. We just don't want to bind them to concrete types in the `types` list.

Let's remove the for loop and within the map have an `if i < implicit_generic_count { fresh var } else { binding }` check

Reply:
> The way you have it also skips over instantiating implicit generics at all, which we don't want

We still instantiate the fresh replacement here I believe? I have switched it to a single map though.